leopard.go

```diff
@@ -0,0 +1,846 @@
+	}
+}
+
+// 4-way butterfly
+func ifftDIT4(work [][]byte, dist int, log_m01, log_m23, log_m02 ffe) {
+	// First layer:
+	if log_m01 == modulus {
+		xor(work[dist], work[0])
+	} else {
+		ifftDIT2(work[0], work[dist], log_m01)
+	}
+
+	if log_m23 == modulus {
+		xor(work[dist*3], work[dist*2])
+	} else {
+		ifftDIT2(work[dist*2], work[dist*3], log_m23)
+	}
+
+	// Second layer:
+	if log_m02 == modulus {
+		xor(work[dist*2], work[0])
+		xor(work[dist*3], work[dist])
+	} else {
+		ifftDIT2(work[0], work[dist*2], log_m02)
+		ifftDIT2(work[dist], work[dist*3], log_m02)
+	}
+}
+
+// 2-way butterfly
+func ifftDIT2(x, y []byte, log_m ffe) {
+	// Reference version:
+	xor(y, x)
+	refMulAdd(x, y, log_m)
+}
+
+// Reference version of muladd: x[] ^= y[] * log_m
+func refMulAdd(x, y []byte, log_m ffe) {
+	lut := mul16LUTs[log_m].LUT
+
+	for off := 0; off < len(x); off += 64 {
+		for i := 0; i < 32; i++ {
+			lo := y[off+i]
+			hi := y[off+i+32]
+
+			prod :=
+				lut[(lo&15)] ^
+					lut[(lo>>4)+16] ^
+					lut[(hi&15)+32] ^
+					lut[(hi>>4)+48]
+
+			x[off+i] ^= byte(prod)
+			x[off+i+32] ^= byte(prod >> 8)
+		}
+	}
+}
+
+func memset(s []byte, n byte) {
+	for i := range s {
+		s[i] = 0
+	}
+}
+
+// xorSlices calls xor for every slice pair in v1, v2.
+func xorSlices(v1, v2 [][]byte) {
+	for i, v := range v1 {
+		xor(v, v2[i])
+	}
+}
+
+// xor sets v1[] = v1[] ^ v2[].
+func xor(v1, v2 []byte) {
+	for i, v := range v1 {
+		v1[i] = v ^ v2[i]
+	}
+}
+
+func mul(x, y []byte, log_m ffe) {
+	refMul(x, y, log_m)
+}
+
+// Reference version of mul: x[] = y[] * log_m
+func refMul(x, y []byte, log_m ffe) {
+	lut := mul16LUTs[log_m].LUT
+
+	for off := 0; off < len(x); off += 64 {
+		for i := 0; i < 32; i++ {
+			lo := y[off+i]
+			hi := y[off+i+32]
+
+			prod :=
+				lut[(lo&15)] ^
+					lut[(lo>>4)+16] ^
+					lut[(hi&15)+32] ^
+					lut[(hi>>4)+48]
+
+			x[off+i] = byte(prod)
+			x[off+i+32] = byte(prod >> 8)
+		}
+	}
+}
+
+// Returns a * Log(b)
+func mulLog(a, log_b ffe) ffe {
+	/*
+	   Note that this operation is not a normal multiplication in a finite
+	   field because the right operand is already a logarithm.  This is done
+	   because it moves K table lookups from the Decode() method into the
+	   initialization step that is less performance critical.  The LogWalsh[]
+	   table below contains precalculated logarithms so it is easier to do
+	   all the other multiplies in that form as well.
+	*/
+	if a == 0 {
+		return 0
+	}
+	return expLUT[addMod(logLUT[a], log_b)]
+}
+
+// z = x + y (mod kModulus)
+func addMod(a, b ffe) ffe {
+	sum := uint(a) + uint(b)
+
+	// Partial reduction step, allowing for kModulus to be returned
+	return ffe(sum + sum>>bitwidth)
+}
+
+// z = x - y (mod kModulus)
+func subMod(a, b ffe) ffe {
+	dif := uint(a) - uint(b)
+
+	// Partial reduction step, allowing for kModulus to be returned
+	return ffe(dif + dif>>bitwidth)
+}
+
+// ceilPow2 returns power of two at or above n.
+func ceilPow2(n int) int {
+	const w = int(unsafe.Sizeof(n) * 8)
```

Comment:
This indeed will be a constant so perhaps we can hoist it out and call it: maxIntBits and have it computed once?

Reply:
I don't understand. I believe all Go constants are computed only once, at compile time?

Reply:
Left it as is. Didn't bother switching to sized `int` type.